index.src.html

```diff
@@ -185,17 +205,43 @@
     When invoked, run these steps:
     1. If |config| is not a <a>valid AudioDecoderConfig</a>, throw a
         {{TypeError}}.
-    2. Run the <a>Configure Decoder</a> algorithm with |config|.
+    2. If {{AudioDecoder/state}} is `"closed"`, throw an {{InvalidStateError}}.
+    3. If the user agent cannot provide a codec implementation to support
+        |config|, throw a {{NotSupportedError}}.
+    4. Set {{AudioDecoder/state}} to `"configured"`.
+    5. <a>Queue a control message</a> to configure the decoder with |config|.
+    6. <a>Run the control message processing loop</a>.
+
+    <a>Running a control message</a> to configure the decoder means running
+    these steps:
+    1. Assign {{AudioDecoder/[[codec implementation]]}} with an implementation
+        supporting |config|.
   </dd>
 
   <dt><dfn method for=AudioDecoder>decode(chunk)</dfn></dt>
   <dd>
     <a>Enqueues a control message</a> to decode the given |chunk|.
 
     When invoked, run these steps:
-    1. Let |output algorithm| be the <a>AudioFrame Output</a> algorithm.
-    2. Run the <a>Decode Chunk</a> algorithm with |chunk| and
-        |output algorithm|.
+    1. If {{AudioDecoder/state}} is not `"configured"`, throw an
+        {{InvalidStateError}}.
+    2. Increment {{AudioDecoder/decodeQueueSize}}.
+    3. <a>Queue a control message</a> to decode the |chunk|.
+    4. <a>Run the control message processing loop</a>.
+
+    Running a control message to decode the chunk means performing these steps:
+    1. Decrement {{AudioDecoder/decodeQueueSize}}
+    2. Let |codec implementation queue| be the result of starting a new
+        <a>parallel queue</a>.
+    3. Enqueue the following steps to |codec implementation queue|:
+        1. Attempt to use {{AudioDecoder/[[codec implementation]]}} to decode
+            the chunk.
+        2. If decoding results in an error, queue a task on the media element
+            task source to run the <a>Close AudioDecoder</a> algorithm with
+            {{EncodingError}}.
+        3. Otherwise, for each decoded audio data |output|, queue a task on the
+            media element task source to run the <a>Output AudioFrame</a>
+            algorithm with |output|.
```

Comment:
I reason like this: there is only one queue, and the size object is counting control messages in that queue where message.type = 'decode'. In our impl, I actually almost implemented it that way (counting messages for every call to the size attribute). 

I didn't follow the second paragraph (why would decoding fail if its just overloaded?).

Reply:
> Say the author has called decode 5 times, but for this particular codec at this time, n is 3. 5 decode jobs are enqueued to the codec parallel queue, decodeQueueSize is now 0, the author think all is good.

This reasoning isn't what I intended. The control message processing loop is run each time the author calls decode(). On the first 3 of those times, the UA finds that it is able to send those packets directly to the underlying codec, so the `decodeQueueSize = 0` after the third call. On the fourth call, the UA finds that the underlying decoder is still saturated, so the 4th decode() control message is not removed from the front of the queue and `decodeQueueSize = 1`. This repeats on the 5th call, leaving `decodeQueueSize = 2`. At some later time the underlying decoder begins to emit outputs from those first 3 packets and this triggers the UA to restart the control processing loop, moving decode 4 and 5 into the underlying decoder. 

> Then the parallel queue starts executing tasks, and the codec implementation quickly errors out because it's overloaded by the 5 jobs.

This would not occur. The UA should not dequeue extra decode() control messages if the underlying decoder doesn't have room. Instead, the queue should start growing, reflected by `decodeQueueSize`.

> but for this you need to start decoding (say, avcodec_send_frame returning EAGAIN, to speak ffmpeg),

In chrome, each decode sent to the underlying decoder is tracked as 'pending' until we here back that its 'done' (which really just means ready for more input) . Upon 'done', we check the control message queue for additional decode() messages to send to the underlying decoder. https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/modules/webcodecs/decoder_template.cc;drc=456bdd17f1d32e10406636109071aec9bf1f1f6e;l=321